### ISSUE TITLE
Perform jwt verification before extauth

### ISSUE THREAD

**Issue by rickducott** [solo-io/gloo#2591]
If I deploy a virtual service that uses both ext auth and jwts, Gloo will order the jwt filter after the ext auth filter. For instance, an envoy http filter chain may look like this: 
```
...
"http_filters": [
             {
              "name": "envoy.fault"
             },
             {
              "name": "envoy.cors"
             },
             {
              "name": "io.solo.filters.http.transformation_ee"
             },
             {
              "name": "io.solo.filters.http.modsecurity",
              "config": {
               "disabled": true
              }
             },
             {
              "name": "io.solo.filters.http.sanitize",
              "config": {
               "headersToRemove": [
                "x-user-id"
               ]
              }
             },
             {
              "config": {
               "grpcService": {
                "timeout": "0.200s",
                "envoyGrpc": {
                 "clusterName": "extauth_gloo-system"
                }
               }
              },
              "name": "envoy.ext_authz"
             },
             {
              "name": "io.solo.filters.http.solo_jwt_authn",
              "config": {
               "filterStateRules": {
                "requires": {},
                "name": "filterState"
               },
               "providers": {}
              }
             },
             {
              "name": "envoy.filters.http.rbac"
             },
             {
              "name": "io.solo.transformation"
             },
             {
              "name": "envoy.rate_limit",
              "config": {
               "requestType": "both",
               "stage": 1,
               "rateLimitService": {
                "grpcService": {
                 "envoyGrpc": {
                  "clusterName": "rate-limit_gloo-system"
                 }
                }
               },
               "domain": "custom",
               "timeout": "0.100s"
              }
             },
             {
              "name": "envoy.rate_limit",
              "config": {
               "requestType": "both",
               "rateLimitService": {
                "grpcService": {
                 "envoyGrpc": {
                  "clusterName": "rate-limit_gloo-system"
                 }
                }
               },
               "domain": "ingress",
               "timeout": "0.100s"
              }
             },
             {
              "name": "io.solo.aws_lambda",
              "config": {
               "useDefaultCredentials": false
              }
             },
             {
              "name": "envoy.router"
             }
...
```

It may be desirable to instead perform the jwt verification in Envoy first, then send the request to the ext auth server. This is not currently configurable. ~In our plugin implementation, we call `jwt` an AuthZ plugin and extauth AuthN, and we order AuthN first.~

It looks like we call `jwt` and `extauth` both `authn` plugins, but don't allow customizing the relative order.

**Comment by rickducott** [comment]
It looks like the order within the the `AuthN` stage is determined by a sort, which reduces to lexicographical ordering.

**Comment by rickducott** [comment]
We can use dynamic metadata for passing context from JWT to extauth (expose as boolean rather than fragile, complex config), so that extauth configs can utilize the JWT.

**Comment by lgadban** [comment]
After thinking through this a bit, if the JWT plugin runs before extauth, wouldn't that remove the ability to do any JWT validation on a token received from an OIDC flow?

I'm not sure if that is actually a concern (i.e. does it even make sense to run JWT validation on a token you just received), but I want to make sure to call out that it would no longer be possible.

**Comment by lgadban** [comment]
Another implication is that now various portions of the OIDC flow will have to disable JWT verification, e.g. the callback URL can't be JWT verified because it doesn't have a JWT yet.

**Comment by srenatus** [comment]
On a related note, using a custom extauth (opa) together with
```yaml
     options:
       headerManipulation:
         requestHeadersToRemove:
           - user-id
           - enterprise
           - tenant
```
we find:

- the request with those headers attached
  - is hitting ext_auth with **those headers present**, 
  - ext_auth will add new headers
  - _those headers_ are not reaching the upstream

suggesting that the requestHeadersToRemove filter is inserted **after extauth**.

👉 Can we affect the ordering there somehow?

**Comment by kdorosh** [comment]
This will be resolved by @saiskee 's current work to add support for stages to our JWT filter, so you can JWT verification both before and after extauth

**Comment by saiskee** [comment]
Work has been completed and available since 1.6.4, PR's are referenced in https://github.com/solo-io/gloo/issues/3207.
